File: 1576-reorder-routes-to-make-all-paths-lead-to-the-city-zero/reorder-routes-to-make-all-paths-lead-to-the-city-zero.java
class Solution {
    private List<List<int[]>> graph;
    private boolean[] visited;
    private int count;

    public int minReorder(int n, int[][] connections) {
        graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        
        for (int[] conn : connections) {
            graph.get(conn[0]).add(new int[]{conn[1], 1}); // Original edge
            graph.get(conn[1]).add(new int[]{conn[0], 0}); // Reverse edge
        }
        
        visited = new boolean[n];
        count = 0;
        
        dfs(0);
        
        return count;
    }
    
    private void dfs(int city) {
        visited[city] = true;
        
        for (int[] neighbor : graph.get(city)) {
            if (!visited[neighbor[0]]) {
                count += neighbor[1]; // If it's an original edge, increment count
                dfs(neighbor[0]);
            }
        }
    }
}

File: 1576-reorder-routes-to-make-all-paths-lead-to-the-city-zero/reorder-routes-to-make-all-paths-lead-to-the-city-zero.c
struct Edge {
    int to;
    int original;
};

void dfs(int city, struct Edge** graph, int* sizes, bool* visited, int* count) {
    visited[city] = true;
    
    for (int i = 0; i < sizes[city]; i++) {
        struct Edge neighbor = graph[city][i];
        if (!visited[neighbor.to]) {
            *count += neighbor.original;
            dfs(neighbor.to, graph, sizes, visited, count);
        }
    }
}

int minReorder(int n, int** connections, int connectionsSize, int* connectionsColSize) {
    struct Edge** graph = (struct Edge**)malloc(n * sizeof(struct Edge*));
    int* sizes = (int*)calloc(n, sizeof(int));
    
    for (int i = 0; i < connectionsSize; i++) {
        sizes[connections[i][0]]++;
        sizes[connections[i][1]]++;
    }
    
    for (int i = 0; i < n; i++) {
        graph[i] = (struct Edge*)malloc(sizes[i] * sizeof(struct Edge));
        sizes[i] = 0;
    }
    
    for (int i = 0; i < connectionsSize; i++) {
        int a = connections[i][0], b = connections[i][1];
        graph[a][sizes[a]++] = (struct Edge){b, 1};
        graph[b][sizes[b]++] = (struct Edge){a, 0};
    }
    
    bool* visited = (bool*)calloc(n, sizeof(bool));
    int count = 0;
    
    dfs(0, graph, sizes, visited, &count);
    
    for (int i = 0; i < n; i++) {
        free(graph[i]);
    }
    free(graph);
    free(sizes);
    free(visited);
    
    return count;
}